simtbx/nanoBragg/nanoBragg_beam.py

```diff
@@ -60,40 +94,83 @@ def polarization_fraction(self, val):
   @property
   def xray_beams(self):
     self._xray_beams = flex_Beam()
+
+    divs = self.divergences
+
+    wavelen = self.spectrum[0][0]
+    nominal_beam = BeamFactory.simple(wavelen * 1e-10)
+    nominal_beam.set_unit_s0(self.unit_s0)
+    nominal_beam.set_polarization_fraction(self.polarization_fraction)
+    beam_vector = nominal_beam.get_sample_to_source_direction()
+    beam_vector /= np.linalg.norm(beam_vector)
+    vert_vector = nominal_beam.get_polarization_normal()
+    polar_vector = np.cross(beam_vector, vert_vector)
+    polar_vector /= np.linalg.norm(polar_vector)
+
+    self.num_div_angles_within_cone = 0
+    beams = []
+    for hdiv, vdiv in divs:
+      vec_xyz = rotate_axis(-beam_vector, polar_vector, vdiv)
+      unit_s0 = rotate_axis(vec_xyz, vert_vector, hdiv)
+      div_ang = np.arccos(np.dot(unit_s0, -beam_vector))
+      if hdiv == 0 and vdiv == 0:
+        assert div_ang == 0
+        assert np.allclose(unit_s0, nominal_beam.get_unit_s0())
+      if div_ang > 1.1*(self.divergence_mrad / 1000. / 2.):
+        continue
```

Comment:
One more comment since I burned myself looking for errors in DIALS recently. Can this condition ever be met? If no, wouldn't `raise ValueError` be more fitting?

Reply:
Aah, I see now. But in this I have two more ideas:

- Isn't 10% a relatively big tolerance? Something like 1e-6 should be completely enough.
- I would strongly suggest baking this condition in self.divergences definition to avoid future errors. If the corner vectors are not physically valid, why ever have them accessible?